Rust/src/common/cluster.rs

```diff
@@ -0,0 +1,338 @@
+use std::cmp::{max, min};
+use rand::{Rng, SeedableRng};
+use rand_chacha::ChaCha20Rng;
+use rayon::iter::{IntoParallelRefMutIterator, ParallelIterator};
+use crate::common::samplesummary::SampleSummary;
+
+const PHASE2_THRESHOLD : usize = 2;
+const separation_ratio_for_merge : f64 = 0.8;
+const weight_threshold : f64 = 1.25;
+const LENGTH_BOUND : usize = 1000;
+
+pub trait Cluster<Q,T: ?Sized> {
+    fn get_weight(&self) -> f64;
+    fn distance_to_point(&self, point : &T, distance: fn(&T, &T) -> f64) -> f64;
+    fn distance_to_cluster(&self, other : &dyn Cluster<Q,T>, distance: fn(&T, &T) -> f64) -> f64;
```

Comment:
Another approach to this in rust could be to define a trait for distance functions:
```
trait DistanceFn<A, B> {
  fn distance(a: A, b: B) ->  f64;
}
```

And then write implementations for the things we want to measure:
```
struct L1Distance {}

impl DistanceFn<f32, f32> for L1Distance {
    fn distance(&self, a: f32, b: f32) -> f64 {
        (a as f64 - b as f64).abs()
    }
}

impl<A, B> DistanceFn<Vec<A>, Vec<B>> for L1Distance
where
    A: Copy,
    B: Copy,
    L1Distance: DistanceFn<A, B>
{
    fn distance(&self, a: Vec<A>, b: Vec<B>) -> f64 {
       a.iter().zip(b.iter()).map(|(x, y)| self.distance(*x, *y)).sum()
    }
}
```

If we take this approach, we might be able to write a few generic impls for distance functions that cover a lot of use cases, and we wouldn't have to add `distance_to_foo` functions to other traits.

Reply:
This approach a possibility -- but this would be the path to erasure of types and dynamic dispatch. (But the clusters themselves are dynamic dispatch already, but that is essential for extensibility). More of these traits can be a slowdown. But more importantly we may not have to solve this issue at this point.

The existence of two distance functions is fundamental. In this case the (&T, &T) -> f64 is the true distance function over immutable references to points. The Cluster however can (and does) a different representation of a point (thought another way, a cluster is a "different type" of an object). So any clustering algorithm should specify the distance to cluster -- as an extension of the point-to-point distance. So that Cluster--to--point is necessary, and so is Cluster--to--Cluster, for every new algorithm/implementation. So the trait definition just forces that clarity. Whether we use traits or just plain objects is probably less of an issue -- but we should probably index on simplicity.

Maybe shelve this and come back to revisit if we are staring at implementing  n x m distances? I expect 3 implementations per algorithm (as described) and no cross products.